text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+
+### Partial Grouping Awareness in Relay
+
+Today our grouping system really can only run in the tail end of the processing pipeline.  This makes it impossible for
+us to perform certain types of filtering or grouping decisions in relay.  There is a range of issues that could already
+be grouped quite accurately in Relay which would then permit us to do metrics extraction in relay or to perform efficent
+discarding.
+
+### Tail-based Sampling / Distributed Sampling Context Buffer
+
+Relay is currently unable to perform tail based sampling or consistent sampling if the dynamic sampling context changes.
+This could be remedied by extensive buffering or alternative approaches.  While a full tail based sampling approach is
+likely to be quite costly, there might be hybrid approaches possible.
+
+## Service Infrastructure
+
+These are changes to the Sentry architecture to better support the creation of independent services.
+
+### In-Monolith Service Compartmentalization
+
+The Sentry monolith currently houses a range of small services that are however clustered together.  For instance the
+processing pipeline really could be separated entirely from the rest of the System.  However the same is true for quite
+a lot of different parts of the product.  For instance certain parts of the user experience are sufficiently idependent
+in the UI code today already that they could turn into a separate service that just supplies UI views together with some
+APIs.  While it's unclear if the existing experiences are worth splitting up into new services, it's probably likely that
+we can at least structure their code so that we can reduce the total amount of pieces of code that need to be checked and
+tested against changes.
+
+In the ideal situation a change to the settings page of a project for instance does not need to run the UI tests for
+performance views etc. unless the settings are in fact related to that component.
+
+### Service Declarations
+
+The biggest limiting factor today in creating new services is the configuration of these services.  When an engineer adds
+a new service there is typically a followup PR against the internal ops repo to configure these services for the main
+sentry installation.  Then there is some sort of ad-hoc workaround to make the whole thing work in local development, how
+to make it work on a single tenant or self hosted installation.  Each new services undergoes a new phase of rediscovery how
+suboptimal the rules and policies around scaling these services are.
+
+Ideally we were able to describe the services in a config file (YAML?) which then can be consumed by deploy tools to
+correctly provision and auto scale services.  This could either be a Sentry specific format or we adopt something that
+has already been tested.
+
+### Serverless Function Declarations for Consumers
+
+The smallest unit of execution at Sentry is not a docker image but in fact a function.  We currently have no real serverless
```

Comment:
related to the rabbitmq vs kafka discussion, but we already have this abstraction in sentry: a celery task

we can consider two almost orthogonal things here:

1. whether we want to continue using celery as our task abstraction, or build something simpler and more restrictive in-house
2. what the backing queue should be (rabbitmq, kafka, something else)